src/toxcore/messenger/packet/file_control.rs

```diff
@@ -0,0 +1,148 @@
+/*! FileControl struct.
+It is used to control transferring file to a friend.
+*/
+
+use nom::{le_u8, be_u64};
+
+use crate::toxcore::binary_io::*;
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+#[repr(u8)]
+/// Whether this node is sender or receiver of file data packet
+pub enum SendReceive {
+    /// This node is sender
+    Send  = 0,
+    /// This node is receiver
+    Receive
+}
+
+impl FromBytes for SendReceive {
+    named!(from_bytes<SendReceive>,
+        switch!(le_u8,
+            0 => value!(SendReceive::Send) |
+            1 => value!(SendReceive::Receive)
+        )
+    );
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+#[repr(u8)]
+/// Control types for transferring file data
+pub enum ControlType {
+    /// Accept a request of transferring file from a peer
+    Accept  = 0,
+    /// Pause transferring
+    Pause,
+    /// Stop transferring and quit session
+    Kill,
+    /// Seek to position of file stream
+    Seek
```

Comment:
`seek_param: Option<u64>` can live in FileControl only if the type is Seek. How about changing `Seek` to `Seek(u64)`? Then you don't have to keep seek_param as option and it will be typesafe to set seek_param **only** if ControlType == Seek.

@kurnevsky ?

Reply:
Yeah, I did the similar thing with even more complex state: https://github.com/tox-rs/tox/blob/5c57ddd68621b2c87a24e181a0134dce0194966f/src/toxcore/net_crypto/crypto_connection.rs#L144